src/Source.jl

```diff
@@ -67,35 +66,35 @@ function sqlitevalue{T}(::Type{T}, handle, col)
 end
 
 # DataStreams interface
-Data.schema(source::SQLite.Source, ::Type{Data.Field}) = source.schema
+Data.schema(source::SQLite.Source) = source.schema
 function Data.isdone(s::Source, row, col)
     (s.status == SQLITE_DONE || s.status == SQLITE_ROW) || sqliteerror(s.stmt.db)
     return s.status == SQLITE_DONE
 end
 # resets an SQLite.Source, ready to read data from at the start of the resultset
-# Data.reset!(io::SQLite.Source) = (sqlite3_reset(io.stmt.handle); execute!(io.stmt))
-Data.streamtype{T<:SQLite.Source}(::Type{T}, ::Type{Data.Field}) = true
+Data.reset!(io::SQLite.Source) = (sqlite3_reset(io.stmt.handle); execute!(io.stmt))
+Data.streamtype(::Type{SQLite.Source}, ::Type{Data.Field}) = true
 
-# `T` might be Int, Float64, String, WeakRefString, any Julia type, Any, NullType
+# `T` might be Int, Float64, String, WeakRefString, any Julia type, Any, Null
 # `t` (the actual type of the value we're returning), might be SQLITE_INTEGER, SQLITE_FLOAT, SQLITE_TEXT, SQLITE_BLOB, SQLITE_NULL
-# `SQLite.streamfrom` returns the next `Nullable{T}` value from the `SQLite.Source`
-function Data.streamfrom{T}(source::SQLite.Source, ::Type{Data.Field}, ::Type{Nullable{T}}, row, col)
+# `SQLite.streamfrom` returns the next `Union{T, Null}` value from the `SQLite.Source`
+function Data.streamfrom(source::SQLite.Source, ::Type{Data.Field}, ::Type{Union{T, Null}}, row, col) where {T}
     handle = source.stmt.handle
     t = SQLite.sqlite3_column_type(handle, col)
     if t == SQLite.SQLITE_NULL
-        val = Nullable{T}()
+        val = null
     else
         TT = SQLite.juliatype(t) # native SQLite Int, Float, and Text types
-        val = Nullable{T}(sqlitevalue(ifelse(TT === Any && !isbits(T), T, TT), handle, col))
+        val = SQLite.sqlitevalue(ifelse(TT === Any && !isbits(T), T, TT), handle, col)
     end
     col == source.schema.cols && (source.status = sqlite3_step(handle))
-    return val
+    return val::Union{T, Null}
 end
-function Data.streamfrom{T}(source::SQLite.Source, ::Type{Data.Field}, ::Type{T}, row, col)
+function Data.streamfrom(source::SQLite.Source, ::Type{Data.Field}, ::Type{T}, row, col) where {T}
     handle = source.stmt.handle
     t = SQLite.sqlite3_column_type(handle, col)
     if t == SQLite.SQLITE_NULL
-        throw(NullException)
+        throw(Data.NullException("encountered null value in non-null typed column: row = $row, col = $col"))
```

Comment:
It would make sense to add a `msg::String` field to `NullException` in Base so that we can unify the exception types.

Reply:
Indeed; I think all Exception types should be able to have a message.